File: android/app/src/main/kotlin/com/swastricare/health/ui/screens/family/member/FamilyNudgeViewModel.kt
package com.swastricare.health.ui.screens.family.member

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.swastricare.health.data.repository.FamilyNudgeRepository
import com.swastricare.health.data.repository.NudgePreset
import com.swastricare.health.data.repository.NudgeResponse
import com.swastricare.health.domain.repository.AuthRepository
import com.swastricare.health.domain.repository.FamilyRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * One-shot UI events emitted after a nudge send completes.
 * Surfaced via the screen's snackbar, then cleared via [FamilyNudgeViewModel.clearEvent]
 * so re-triggering works without stale recomposition.
 */
sealed class NudgeUiEvent {
    /**
     * Send succeeded. [deliveredViaPush] mirrors the edge function's `delivered`
     * field (true → FCM accepted, false → recipient offline / no token / opted out).
     * [reason] is the edge function's diagnostic ("no_token", "quiet_hours", …) when
     * `delivered=false`.
     */
    data class Success(val deliveredViaPush: Boolean, val reason: String?) : NudgeUiEvent()

    /** Send failed locally (network, missing recipient, validation, …). */
    data class Failure(val message: String) : NudgeUiEvent()
}

/**
 * UI state for the family nudge bottom sheet.
 *
 * [recipientUserId] is null until [FamilyNudgeViewModel.init] resolves the target
 * member from the caller's family group; the screen disables send actions while null.
 */
data class FamilyNudgeState(
    val memberName: String = "",
    val recipientUserId: String? = null,
    val targetHealthProfileId: String = "",
    val customMessage: String = "",
    val isSending: Boolean = false,
    val event: NudgeUiEvent? = null,
)

/**
 * Hosts the nudge bottom sheet's state + send actions.
 *
 * Resolves the target [com.swastricare.health.domain.model.FamilyMember] from the
 * caller's own family group (via [FamilyRepository.getMyFamilyGroup] +
 * [FamilyRepository.getMembers]) so we know the recipient's `auth.users.id` for
 * the FCM fan-out — `healthProfileId` alone is insufficient since edge fn needs userId.
 */
@HiltViewModel
class FamilyNudgeViewModel @Inject constructor(
    private val familyRepository: FamilyRepository,
    private val authRepository: AuthRepository,
    private val nudgeRepository: FamilyNudgeRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(FamilyNudgeState())
    val state: StateFlow<FamilyNudgeState> = _state.asStateFlow()

    fun init(targetHealthProfileId: String) {
        _state.value = _state.value.copy(targetHealthProfileId = targetHealthProfileId)
        viewModelScope.launch {
            runCatching {
                val callerUserId = authRepository.getCurrentUser()?.id ?: return@launch
                val group = familyRepository.getMyFamilyGroup(callerUserId).getOrNull()
                    ?: return@launch
                val members = familyRepository.getMembers(group.id).getOrNull().orEmpty()
                val target = members.firstOrNull { it.healthProfileId == targetHealthProfileId }
                _state.value = _state.value.copy(
                    memberName = target?.fullName.orEmpty(),
                    recipientUserId = target?.userId,
                )
            }
        }
    }

    fun setCustomMessage(text: String) {
        _state.value = _state.value.copy(customMessage = text.take(200))
    }

    fun sendPreset(preset: NudgePreset) {
        val recipient = _state.value.recipientUserId ?: run {
            emit(NudgeUiEvent.Failure("Member not loaded"))
            return
        }
        send {
            nudgeRepository.sendPreset(recipient, _state.value.targetHealthProfileId, preset)
        }
    }

    fun sendCustom() {
        val recipient = _state.value.recipientUserId ?: run {
            emit(NudgeUiEvent.Failure("Member not loaded"))
            return
        }
        val msg = _state.value.customMessage.trim()
        if (msg.isEmpty()) {
            emit(NudgeUiEvent.Failure("Message cannot be empty"))
            return
        }
        send {
            nudgeRepository.sendCustom(recipient, _state.value.targetHealthProfileId, msg)
        }
    }

    private fun send(block: suspend () -> Result<NudgeResponse>) {
        _state.value = _state.value.copy(isSending = true, event = null)
        viewModelScope.launch {
            val r = block()
            _state.value = _state.value.copy(isSending = false)
            r.fold(
                onSuccess = { resp ->
                    emit(NudgeUiEvent.Success(deliveredViaPush = resp.delivered, reason = resp.reason))
                    _state.value = _state.value.copy(customMessage = "")
                },
                onFailure = { e -> emit(NudgeUiEvent.Failure(e.message ?: "Send failed")) },
            )
        }
    }

    fun clearEvent() {
        _state.value = _state.value.copy(event = null)
    }

    private fun emit(event: NudgeUiEvent) {
        _state.value = _state.value.copy(event = event)
    }
}

File: swastricare-mobile-swift/ViewModels/Family/FamilyNudgeViewModel.swift
//
//  FamilyNudgeViewModel.swift
//  swastricare-mobile-swift
//
//  Family Monitoring — ViewModel powering the nudge bottom sheet.
//  Resolves the recipient's user id from their health profile, then sends
//  preset or custom nudges via FamilyNudgeRepository.
//
//  Mirrors the Android FamilyNudgeViewModel.
//

import Foundation
import SwiftUI
import Combine
import Supabase

@MainActor
final class FamilyNudgeViewModel: ObservableObject {

    // MARK: - Event

    enum Event {
        case success(deliveredViaPush: Bool, reason: String?)
        case failure(message: String)
    }

    // MARK: - State

    struct State {
        var memberName: String = ""
        var recipientUserId: String?
        var targetHealthProfileId: String = ""
        var customMessage: String = ""
        var isSending: Bool = false
        var event: Event?
    }

    @Published var state = State()

    // MARK: - Dependencies

    private let supabase = SupabaseManager.shared
    private let repo = FamilyNudgeRepository.shared

    // MARK: - Init

    /// Resolves the target family member (name + owner user id) so we can send
    /// nudges. Uses the same family group lookup flow as the dashboard.
    func initialize(targetHealthProfileId: String) async {
        state.targetHealthProfileId = targetHealthProfileId

        do {
            // 1. Find the caller's family group, then the target member row
            //    (FamilyMember carries fullName via embedded health_profiles).
            if let group = try await supabase.fetchMyFamilyGroup() {
                let members = try await supabase.fetchFamilyMembers(groupId: group.id)
                if let target = members.first(where: {
                    $0.healthProfileId.uuidString.lowercased() == targetHealthProfileId.lowercased()
                }) {
                    state.memberName = target.fullName ?? ""
                }
            }

            // 2. Resolve the recipient's auth user id by looking up health_profiles
            //    directly (FamilyMember doesn't carry user_id).
            struct ProfileRow: Decodable { let user_id: String? }
            let rows: [ProfileRow] = try await supabase.client
                .from("health_profiles")
                .select("user_id")
                .eq("id", value: targetHealthProfileId)
                .limit(1)
                .execute()
                .value
            if let uid = rows.first?.user_id, !uid.isEmpty {
                state.recipientUserId = uid
            }
        } catch {
            // Leave recipientUserId nil → sheet shows "Resolving member…" hint.
            // Don't surface as an Event because there's nothing the user can do.
        }
    }

    // MARK: - Inputs

    func setCustomMessage(_ text: String) {
        state.customMessage = String(text.prefix(200))
    }

    func clearEvent() {
        state.event = nil
    }

    // MARK: - Send

    func sendPreset(_ preset: NudgePreset) {
        guard let recipient = state.recipientUserId else {
            state.event = .failure(message: "Member not loaded")
            return
        }
        state.isSending = true
        state.event = nil
        Task {
            do {
                let resp = try await repo.sendPreset(
                    recipientUserId: recipient,
                    targetProfileId: state.targetHealthProfileId,
                    preset: preset,
                    isCritical: false
                )
                state.isSending = false
                state.event = .success(deliveredViaPush: resp.delivered, reason: resp.reason)
                state.customMessage = ""
            } catch {
                state.isSending = false
                state.event = .failure(message: error.localizedDescription)
            }
        }
    }

    func sendCustom() {
        let msg = state.customMessage.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !msg.isEmpty else {
            state.event = .failure(message: "Message cannot be empty")
            return
        }
        guard let recipient = state.recipientUserId else {
            state.event = .failure(message: "Member not loaded")
            return
        }
        state.isSending = true
        state.event = nil
        Task {
            do {
                let resp = try await repo.sendCustom(
                    recipientUserId: recipient,
                    targetProfileId: state.targetHealthProfileId,
                    message: msg,
                    category: "CHECKIN",
                    isCritical: false
                )
                state.isSending = false
                state.event = .success(deliveredViaPush: resp.delivered, reason: resp.reason)
                state.customMessage = ""
            } catch {
                state.isSending = false
                state.event = .failure(message: error.localizedDescription)
            }
        }
    }
}
